docs/proposals/deployment-hooks.md

```diff
@@ -0,0 +1,308 @@
+
+If a new deployment is started in middle of the rollout or while the deployment lifecycle
+hook is being executed the lifecycle Pod will be aborted and the new rollout will be
+triggered. Note that this might lead in data corruption if the hooks are used to migrate
+application data to database and users are advised not mutate the deployment spec while
+another deployment is making progress.
+
+Every failure of a lifecycle hook should be logged as an Event regardless of the failure
+policy.
+
+### Default values
+
+Initially the default `FailurePolicy` should be set to `Retry` until the deployments can
+safely be rolled back automatically.
+
+### Associating lifecycle Pods to Deployment
+
+The `OwnerReference` should be considered to be set on lifecycle Pods to a ReplicaSet that
+represents the deployment they were created for. This allows to garbage collect when the
+ReplicaSet is deleted.
+
+### Garbage Collection
+
+After a successfull deployment the lifecycle hook pods should be considered for garbage
+collection. It it questionable if the failed lifecycle hooks should be opted-out from
+garbage collection to allow future inspection.
+
+## API
+
+The following structures will be added:
+
+```go
+type LifecycleHook struct {
+  // FailurePolicy specifies what action to take if the hook fails.
+  FailurePolicy LifecycleHookFailurePolicy
+
+  // ExecNewPod specifies the options for a lifecycle hook backed by a pod.
+  ExecNewPod *ExecNewPodHook
+}
+```
+
+```go
+type LifecycleHookFailurePolicy string
+
+const (
+  // LifecycleHookFailurePolicyRetry means retry the hook until it succeeds.
+  LifecycleHookFailurePolicyRetry LifecycleHookFailurePolicy = "Retry"
+
+  // LifecycleHookFailurePolicyAbort means abort the deployment (if possible).
+  LifecycleHookFailurePolicyAbort LifecycleHookFailurePolicy = "Abort"
+
+  // LifecycleHookFailurePolicyIgnore means ignore failure and continue the deployment.
+  LifecycleHookFailurePolicyIgnore LifecycleHookFailurePolicy = "Ignore"
+)
+```
+
+```go
+type ExecNewPodHook struct {
+  // Command specifies the ENTRYPOINT of the container image
+  Command []string
+
+  // Args specifies the CMD of the container image
+  Args []string
+
+  // Env is a set of environment variables to supply to the hook pod's container.
+  Env []api.EnvVar
+
+  // ContainerName is the name of a container in the deployment pod template
+  // whose Docker image will be used for the hook pod's container.
+  ContainerName string
+
+  // Volumes is a list of named volumes from the pod template which should be
+  // copied to the hook pod. Volumes names not found in pod spec are ignored.
+  // An empty list means no volumes will be copied.
+  Volumes []string
+}
+```
+
+The `RollingUpdateDeployment` API object will add these fields:
+
+```go
+type RollingUpdateDeployment struct {
+  // TimeoutSeconds is the time to wait for updates before giving up. If the
+  // value is nil, a default will be used.
+  TimeoutSeconds *int64
```

Comment:
No need for this, ProgressDeadlineSeconds will be in the spec. You should mention that this proposal is depending on the perma-failed PR.